scripts/gen_mmu.py

```diff
@@ -0,0 +1,425 @@
+import os
+import sys
+import struct
+import parser
+from collections import namedtuple
+import ctypes
+import argparse
+
+############# global variables
+pd_complete = ''
+inputfile = ''
+outputfile = ''
+list_of_pde = {}
+num_of_regions = 0
+read_buff=''
+
+struct_mmu_regions_tuple = {"start_addr","size","permissions"}
+mmu_region_details = namedtuple("mmu_region_details", "start_addr size pde_index \
+                                pte_valid_addr_start \
+                                pte_valid_addr_end \
+                                permissions")
+
+page_tables_list = []
+PD_start_addr = 0
+validation_issue_memory_overlap = False
+output_offset = 0
+
+############# templates
+basic_header = """
+#include <mmustructs.h>
+"""
+
+
+template_pd_complete = """
+static struct x86_mmu_page_directory page_directory = {
+REPACEMENT_HOOK
+};
+"""
+template_disabled_pde = """entry[PDE_NUMBER].pt = {.value = 0 }"""
+
+template_enabled_pde = """    .entry[PDE_NUMBER].pt = {
+        .p = 1,
+        .rw = PERMISSIONS,
+        .us = SUPERVISOR,
+        .pwt = 0/*TODO*/,
+        .pcd = 0/*TODO*/,
+        .a = 0,
+        .ps = 0 ,
+        .page_table = ADDRESS_OF_PAGE_TABLE},\n"""
+
+template_pt_complete = """
+__attribute__((aligned(4096))) static struct x86_mmu_page_table page_table_PDE_NUMBER = {
+REPACEMENT_HOOK
+};
+"""
+template_disabled_pte = """entry[PTE_NUMBER] = {.value = 0 }"""
+
+template_enabled_pte = """    .entry[PTE_NUMBER] = {
+        .p = 1,
+        .rw = PERMISSIONS,
+        .us = SUPERVISOR,
+        .pwt = 0/*TODO*/,
+        .pcd = 0/*TODO*/,
+        .a = 0,
+        .d = 0,
+        .pat = 0,
+        .g = 0,
+        .alloc = 1 ,
+        .custom = 0,
+        .page = ADDRESS_OF_PHYSICAL_PAGE},\n"""
+#############
+
+#return the page directory number for the give address
+def get_pde_number(value):
+    return( (value >> 22 ) & 0x3FF)
+
+#return the page table number for the given address
+def get_pte_number(value):
+    return( (value >> 12 ) & 0x3FF)
+
+
+#consider little endian
+def read_little_endian_binay(file_name):
+    actual_word = [0,0,0,0]
+    cal_number = 0
+    for i in range(4):
+        cal_number += (int(ord(file_name.read(1)))  << (8*i))
+    return cal_number
+
+# read the binay from the input file and populate a dict for
+# start address of mem region
+# size of the region - so page tables entries will be created with this
+# read write permissions
+def read_mmu_list_marshal_param():
+
+    global read_buff
+    global page_tables_list
+    global PD_start_addr
+    global validation_issue_memory_overlap
```

Comment:
get rid of this variable and just throw an exception as soon as a problem is detected